File: deployment/backend/app/tracker.py
from dataclasses import dataclass

from app.config import (
    ALARM_WARNING_THRESHOLD,
    CONTINUOUS_DROWSY_THRESHOLD,
    SUSPECT_THRESHOLD,
    WARNING_THRESHOLD,
)


@dataclass
class TrackerState:
    status: str
    drowsy_counter: int
    warning_counter: int
    alarm_active: bool


@dataclass
class DrowsinessTracker:
    drowsy_counter: int = 0
    warning_counter: int = 0
    is_warning_active: bool = False

    def update(self, detected_label: str) -> TrackerState:
        if detected_label == "drowsy":
            self.drowsy_counter += 1
        else:
            self.drowsy_counter = 0
            self.is_warning_active = False

        self.drowsy_counter = min(
            self.drowsy_counter,
            CONTINUOUS_DROWSY_THRESHOLD,
        )

        if self.drowsy_counter >= WARNING_THRESHOLD:
            status = "WARNING!"

            if not self.is_warning_active:
                self.warning_counter += 1
                self.is_warning_active = True
        elif self.drowsy_counter >= SUSPECT_THRESHOLD:
            status = "Suspected Drowsy"
        else:
            status = "Awake"

        alarm_active = (
            self.warning_counter >= ALARM_WARNING_THRESHOLD
            or self.drowsy_counter >= CONTINUOUS_DROWSY_THRESHOLD
        )

        if alarm_active:
            self.reset()

        return TrackerState(
            status=status,
            drowsy_counter=self.drowsy_counter,
            warning_counter=self.warning_counter,
            alarm_active=alarm_active,
        )

    def reset(self) -> None:
        self.drowsy_counter = 0
        self.warning_counter = 0
        self.is_warning_active = False

File: deployment/frontend/src/tracker.js
const SUSPECT_THRESHOLD = 5;
const WARNING_THRESHOLD = 10;
const ALARM_WARNING_THRESHOLD = 3;
export const CONTINUOUS_DROWSY_THRESHOLD = 20;

export function createTracker() {
  let drowsyCounter = 0;
  let warningCounter = 0;
  let isWarningActive = false;

  function reset() {
    drowsyCounter = 0;
    warningCounter = 0;
    isWarningActive = false;
  }

  function update(detectedLabel) {
    if (detectedLabel === "drowsy") {
      drowsyCounter += 1;
    } else {
      drowsyCounter = 0;
      isWarningActive = false;
    }

    drowsyCounter = Math.min(drowsyCounter, CONTINUOUS_DROWSY_THRESHOLD);

    let status = "Awake";
    if (drowsyCounter >= WARNING_THRESHOLD) {
      status = "WARNING!";

      if (!isWarningActive) {
        warningCounter += 1;
        isWarningActive = true;
      }
    } else if (drowsyCounter >= SUSPECT_THRESHOLD) {
      status = "Suspected Drowsy";
    }

    const alarmActive =
      warningCounter >= ALARM_WARNING_THRESHOLD ||
      drowsyCounter >= CONTINUOUS_DROWSY_THRESHOLD;

    if (alarmActive) {
      reset();
    }

    const state = {
      status,
      drowsy_counter: drowsyCounter,
      warning_counter: warningCounter,
      alarm_active: alarmActive,
    };

    return state;
  }

  return {
    update,
    reset,
  };
}
